paddle/fluid/memory/allocation/allocator.cc

```diff
@@ -27,16 +27,24 @@ bool Allocator::IsAllocThreadSafe() const { return false; }
 
 AllocationPtr Allocator::Allocate(size_t size, Allocator::Attr attr) {
   auto ptr = AllocateImpl(size, attr);
-  ptr->set_allocator(this);
+  ptr->RegisterDecoratedAllocator(this);
   return AllocationPtr(ptr);
 }
 
-void Allocator::Free(Allocation* allocation) { delete allocation; }
+void Allocator::FreeImpl(Allocation* allocation) {
+  Allocator* allocator = allocation->TopDecoratedAllocator();
+  allocator->Free(allocation);
+}
+
+void Allocator::Free(Allocation* allocation) {
+  allocation->PopDecoratedAllocator();
```

Comment:
is this poping the last allocator? what not pop in FreeImpl

Reply:
`FreeImpl` is a virtual function that must be overwritten by derived class. If we pop allocator in `FreeImpl`, each derived class must write this code in their implemented function. I just want to hide the implementation of `decorated_allocators_` in base class.